File: scripts/prepare_read_tags.py
import argparse
import pandas as pd


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--input", required=True, help="输入表格，支持 .tsv/.csv")
    parser.add_argument("--output", required=True, help="输出 TSV")
    parser.add_argument("--read-id-col", default="read_id")
    parser.add_argument("--cell-col", default="cell_id")
    parser.add_argument("--umi-primary-col", default="putative_umi_5p")
    parser.add_argument("--umi-backup-col", default="putative_umi")
    parser.add_argument("--barcode-5p-col", default="BC5n")
    parser.add_argument("--barcode-3p-col", default="BC3n")
    return parser.parse_args()


def read_table(path):
    if path.endswith(".csv"):
        return pd.read_csv(path)
    return pd.read_csv(path, sep="\t")


def main():
    args = parse_args()
    df = read_table(args.input).copy()

    required = [
        args.read_id_col,
        args.cell_col,
        args.umi_primary_col,
        args.umi_backup_col,
        args.barcode_5p_col,
        args.barcode_3p_col,
    ]
    missing = [c for c in required if c not in df.columns]
    if missing:
        raise ValueError(f"缺少列: {missing}")

    df[args.umi_primary_col] = df[args.umi_primary_col].fillna("").astype(str).str.strip()
    df[args.umi_backup_col] = df[args.umi_backup_col].fillna("").astype(str).str.strip()
    df[args.barcode_5p_col] = df[args.barcode_5p_col].fillna("").astype(str).str.strip()
    df[args.barcode_3p_col] = df[args.barcode_3p_col].fillna("").astype(str).str.strip()
    df[args.read_id_col] = df[args.read_id_col].astype(str).str.strip()
    df[args.cell_col] = df[args.cell_col].astype(str).str.strip()

    # Strint2.1 keeps both barcode ends, while still emitting Sockeye-style
    # single CB/CR/UR fields for downstream BAM tag tools.
    df["umi_for_clustering"] = df[args.umi_primary_col]
    mask_empty = df["umi_for_clustering"].eq("")
    df.loc[mask_empty, "umi_for_clustering"] = df.loc[mask_empty, args.umi_backup_col]
    df["barcode_dual"] = df[args.barcode_5p_col] + "+" + df[args.barcode_3p_col]
    df.loc[df[args.barcode_5p_col].eq(""), "barcode_dual"] = "+" + df.loc[
        df[args.barcode_5p_col].eq(""), args.barcode_3p_col
    ]
    df.loc[df[args.barcode_3p_col].eq(""), "barcode_dual"] = df.loc[
        df[args.barcode_3p_col].eq(""), args.barcode_5p_col
    ] + "+"

    out = df[
        [
            args.read_id_col,
            args.cell_col,
            args.barcode_5p_col,
            args.barcode_3p_col,
            "barcode_dual",
            args.umi_primary_col,
            args.umi_backup_col,
            "umi_for_clustering",
        ]
    ].copy()
    out.columns = [
        "read_id",
        "cell_id",
        "barcode_5p",
        "barcode_3p",
        "barcode_dual",
        "umi_primary",
        "umi_backup",
        "umi_for_clustering",
    ]

    out = out[out["read_id"] != ""]
    out = out[out["cell_id"] != ""]
    out = out[out["umi_for_clustering"] != ""]

    # 如果同一个 read_id 有重复记录，保留第一条
    out = out.drop_duplicates(subset=["read_id"], keep="first")

    out.to_csv(args.output, sep="\t", index=False)


if __name__ == "__main__":
    main()

File: src/bin/prepare_read_tags.rs
use std::collections::HashSet;
use std::path::PathBuf;

use anyhow::{Context, Result};
use clap::Parser;

#[derive(Debug, Parser)]
#[command(version, about = "Prepare Sockeye-style read tag table from Flora assigned reads")]
struct Cli {
    #[arg(long)]
    input: PathBuf,

    #[arg(long)]
    output: PathBuf,

    #[arg(long = "read-id-col", default_value = "read_id")]
    read_id_col: String,

    #[arg(long = "cell-col", default_value = "cell_id")]
    cell_col: String,

    #[arg(long = "umi-primary-col", default_value = "putative_umi_5p")]
    umi_primary_col: String,

    #[arg(long = "umi-backup-col", default_value = "putative_umi")]
    umi_backup_col: String,

    #[arg(long = "barcode-5p-col", default_value = "BC5n")]
    barcode_5p_col: String,

    #[arg(long = "barcode-3p-col", default_value = "BC3n")]
    barcode_3p_col: String,
}

fn main() -> Result<()> {
    let cli = Cli::parse();
    run(cli)
}

fn run(cli: Cli) -> Result<()> {
    let delimiter = if cli.input.extension().and_then(|x| x.to_str()) == Some("csv") {
        b','
    } else {
        b'\t'
    };
    let mut reader = csv::ReaderBuilder::new()
        .delimiter(delimiter)
        .from_path(&cli.input)
        .with_context(|| format!("read {}", cli.input.display()))?;
    let headers = reader.headers()?.clone();

    let read_idx = required_idx(&headers, &cli.read_id_col)?;
    let cell_idx = required_idx(&headers, &cli.cell_col)?;
    let umi_primary_idx = required_idx(&headers, &cli.umi_primary_col)?;
    let umi_backup_idx = required_idx(&headers, &cli.umi_backup_col)?;
    let barcode_5p_idx = required_idx(&headers, &cli.barcode_5p_col)?;
    let barcode_3p_idx = required_idx(&headers, &cli.barcode_3p_col)?;

    let mut writer = csv::WriterBuilder::new()
        .delimiter(b'\t')
        .from_path(&cli.output)
        .with_context(|| format!("write {}", cli.output.display()))?;
    writer.write_record([
        "read_id",
        "cell_id",
        "barcode_5p",
        "barcode_3p",
        "barcode_dual",
        "umi_primary",
        "umi_backup",
        "umi_for_clustering",
    ])?;

    let mut seen = HashSet::new();
    for record in reader.records() {
        let record = record?;
        let read_id = clean(record.get(read_idx));
        let cell_id = clean(record.get(cell_idx));
        let barcode_5p = clean(record.get(barcode_5p_idx));
        let barcode_3p = clean(record.get(barcode_3p_idx));
        let umi_primary = clean(record.get(umi_primary_idx));
        let umi_backup = clean(record.get(umi_backup_idx));
        let umi_for_clustering = if umi_primary.is_empty() {
            umi_backup.clone()
        } else {
            umi_primary.clone()
        };
        if read_id.is_empty() || cell_id.is_empty() || umi_for_clustering.is_empty() {
            continue;
        }
        if !seen.insert(read_id.clone()) {
            continue;
        }
        let barcode_dual = format!("{barcode_5p}+{barcode_3p}");
        writer.write_record([
            read_id,
            cell_id,
            barcode_5p,
            barcode_3p,
            barcode_dual,
            umi_primary,
            umi_backup,
            umi_for_clustering,
        ])?;
    }
    writer.flush()?;
    Ok(())
}

fn required_idx(headers: &csv::StringRecord, name: &str) -> Result<usize> {
    headers
        .iter()
        .position(|header| header == name)
        .with_context(|| format!("missing required column: {name}"))
}

fn clean(value: Option<&str>) -> String {
    value.unwrap_or("").trim().to_string()
}
